File: l33tcode/compare-version-numbers.py
class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:
        ptr1, ptr2 = 0, 0

        while ptr1 < len(version1) or ptr2 < len(version2):
            num1, num2 = 0, 0

            while ptr1 < len(version1) and version1[ptr1] != ".":
                num1 = num1 * 10 + int(version1[ptr1])
                ptr1 += 1
            ptr1 += 1

            while ptr2 < len(version2) and version2[ptr2] != ".":
                num2 = num2 * 10 + int(version2[ptr2])
                ptr2 += 1
            ptr2 += 1

            if num1 < num2:
                return -1
            elif num1 > num2:
                return 1

        return 0

File: l33tcode_rs/src/compare_version_numbers.rs
#[allow(dead_code)]
pub struct Solution;

impl Solution {
    #[allow(dead_code, clippy::needless_pass_by_value)]
    pub fn compare_version(version1: String, version2: String) -> i32 {
        let mut iter1 = version1.split('.').map(|s| s.parse::<i32>().unwrap());
        let mut iter2 = version2.split('.').map(|s| s.parse::<i32>().unwrap());

        loop {
            let v1 = iter1.next();
            let v2 = iter2.next();

            match (v1, v2) {
                (None, None) => return 0,
                (None, Some(n)) => {
                    if n > 0 {
                        return -1;
                    }
                }
                (Some(n), None) => {
                    if n > 0 {
                        return 1;
                    }
                }
                (Some(n1), Some(n2)) => {
                    if n1 < n2 {
                        return -1;
                    }
                    if n1 > n2 {
                        return 1;
                    }
                }
            }
        }
    }
}
